nx/lib/nx/defn/grad.ex

```
@@ -73,198 +67,198 @@ defmodule Nx.Defn.Grad do
+    case nodes do
+      %{^id => _} -> {parents, nodes}
+      %{} -> parents_args(op, t, id, {parents, Map.put(nodes, id, t)})
     end
   end
 
-  defp zerofy_each(%T{data: %Expr{op: :metadata, args: [t, %{__MODULE__ => code}]}}, cache, ids) do
-    case code do
-      {:sum, exprs} ->
-        {exprs, cache} =
-          Enum.map_reduce(exprs, cache, fn expr, cache ->
-            {expr, cache, _} = zerofy_ids(expr, cache, ids)
-            {expr, cache}
-          end)
+  defp parents_args(:metadata, %{data: %{args: [_, %{stop_grad: true}]}}, _id, acc) do
+    acc
+  end
 
-        {Enum.reduce(exprs, &Nx.add/2), cache, true}
+  # We register syntax nodes in a special slot and make sure they are traversed
+  # early on. We could track the nodes themselves, but this is conceptually simpler.
+  defp parents_args(op, _, id, {parents, nodes}) when op in @syntax do
+    {Map.update(parents, __MODULE__, [id], &[id | &1]), nodes}
+  end
 
-      {:tainted, id} ->
-        if Map.has_key?(ids, id) do
-          {Expr.tensor(0.0), cache, true}
+  defp parents_args(op, t, parent_id, acc) do
+    {_, acc} =
+      traverse_args(op, t, acc, fn %T{data: %{id: id, op: op}} = arg, {parents, nodes} ->
+        if op in @constants do
+          {arg, {parents, nodes}}
         else
-          {t, cache, false}
+          parents = Map.update(parents, id, [parent_id], &[parent_id | &1])
+          {arg, recur_parents_tree(arg, {parents, nodes})}
         end
+      end)
 
-      :to_grad ->
-        {t, cache, false}
-    end
+    acc
   end
 
-  defp zerofy_each(t, cache, ids) do
-    {args, {cache, tainted?}} =
-      Tree.apply_args(t, {cache, false}, fn arg, {cache, acc_tainted?} ->
-        {arg, cache, tainted?} = zerofy_ids(arg, cache, ids)
-        {arg, {cache, tainted? or acc_tainted?}}
-      end)
+  # For some functions, only a subset of the args participate in the grad,
+  # so we handle them accordingly here.
```

How can we ensure new grad implementations pass through here as needed?

Perhaps we explicitly add all functions to either an specific clause or to a clause guard `when fun in ~w(a b c d...)a` instead of a catch-all clause.

This would make it so a new grad raises here and pulls the developers attention to ensure the implementation takes this part into account.

---

I added an automatic check that runs in our tests only. :)